hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/MarkerHandler.java

```diff
@@ -0,0 +1,297 @@
+      }
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to read MARKERS file " + markerDirPath, e);
+    }
+    return markers;
+  }
+
+  private class CreateMarkerCompletableFuture extends CompletableFuture<String> {
+    private Context context;
+    private final String markerDirPath;
+    private final String markerName;
+    private boolean result;
+    private final long startTimeMs;
+
+    public CreateMarkerCompletableFuture(Context context, String markerDirPath, String markerName) {
+      super();
+      this.startTimeMs = System.currentTimeMillis();
+      this.context = context;
+      this.markerDirPath = markerDirPath;
+      this.markerName = markerName;
+      this.result = false;
+    }
+
+    public Context getContext() {
+      return context;
+    }
+
+    public String getMarkerDirPath() {
+      return markerDirPath;
+    }
+
+    public String getMarkerName() {
+      return markerName;
+    }
+
+    public boolean getResult() {
+      return result;
+    }
+
+    public void setResult(boolean result) {
+      LOG.info("*** Request queued for " + (System.currentTimeMillis() - startTimeMs) + " ms");
+      this.result = result;
+    }
+  }
+
+  private class BatchCreateMarkerRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      LOG.info("Start processing create marker requests");
+      long startTimeMs = System.currentTimeMillis();
+      List<CreateMarkerCompletableFuture> futuresToRemove = new ArrayList<>();
+      Set<String> updatedMarkerDirPaths = new HashSet<>();
+      int markerFileIndex = -1;
+      synchronized (isMarkerFileInUseList) {
+        for (int i = 0; i < isMarkerFileInUseList.size(); i++) {
+          if (!isMarkerFileInUseList.get(i)) {
+            markerFileIndex = i;
+            isMarkerFileInUseList.set(i, true);
+            break;
+          }
+        }
+        if (markerFileIndex < 0) {
+          LOG.info("All marker files are busy, skip batch processing of create marker requests in " + (System.currentTimeMillis() - startTimeMs) + " ms");
+          return;
+        }
+      }
+
+      LOG.info("timeMs=" + System.currentTimeMillis() + " markerFileIndex=" + markerFileIndex);
+      synchronized (createMarkerRequestlockObject) {
+        LOG.info("Iterating through existing requests, size=" + createMarkerFutures.size());
+        for (CreateMarkerCompletableFuture future : createMarkerFutures) {
+          String markerDirPath = future.getMarkerDirPath();
+          String markerName = future.getMarkerName();
+          LOG.info("--- markerDirPath=" + markerDirPath + " markerName=" + markerName);
+          Set<String> allMarkers = allMarkersMap.computeIfAbsent(markerDirPath, k -> new HashSet<>());
+          StringBuilder stringBuilder = fileMarkersMap.computeIfAbsent(markerDirPath, k -> new HashMap<>())
+              .computeIfAbsent(markerFileIndex, k -> new StringBuilder(16384));
```

Comment:
what happens if it grows beyond the allocated size?

Reply:
This size, `16384`, is the initial allocated size of the `StringBuilder`.  Once it grows beyond the allocated size, internally the `StringBuilder` allocates new memory space with a larger size (2x) and copies the String content over.